Improve Floppy Controller

chschnell opened the pull request:
This PR fixes the issues collected in #1362:

* #1003
* #1063
* #1358
* #1050
* missing second floppy drive `fdb`
* missing support for unformatted, blank disk images

There are two TODO comments left (corner-cases).

**Known issues**

- **Win98SE:** It is [reported](https://github.com/copy/v86/issues/1362#issuecomment-3131199190) that it's neccessary to remove and re-install the Floppy Disk Controller device in pre-existing Win98SE installations, or else they will block when accessing floppy devices.
- **Linux:** hwinfo reports read errors when no floppy disk is inserted into a drive

copy commented:
> What is the use case for this auto-grow feature? 

IIRC, I found some OSes with build systems that produce a truncated floppy disk file. Not sure if that was only boot sector style images (which aren't really floppy disks), or also 1.44 images. I think keeping it for 512 byte images would be nice, and removing it for the rest is fine. I don't think 512 byte images should be supported by the `set_fda()` mechanism anyway.

> I want to incorporate the disk format detection from qemu

Just a note, qemu's license isn't compatible with v86's, so please don't copy-paste significant portions of code from qemu.

chschnell commented:
> > What is the use case for this auto-grow feature?
> 
> IIRC, I found some OSes with build systems that produce a truncated floppy disk file. Not sure if that was only boot sector style images (which aren't really floppy disks), or also 1.44 images.

Maybe they were one of the many sizes below 1.44MB that I listed? I wasn't even aware that there are that many floppy disk formats in the field until I found that list in qemu.

> I think keeping it for 512 byte images would be nice, and removing it for the rest is fine. I don't think 512 byte images should be supported by the `set_fda()` mechanism anyway.

Yes, the 512 byte "bootsector" images are a special case, they are not regular disks. How else do you consider to support them?

> > I want to incorporate the disk format detection from qemu
> 
> Just a note, qemu's license isn't compatible with v86's, so please don't copy-paste significant portions of code from qemu.

Oh, I wasn't aware of this at all.

In quite a few cases I changed all variable and function names, but the algorithms are mostly the same. I'll have to think about that. What exactly is the license issue?

chschnell commented:
> Just curious, will a `set_fdb` and `eject_fdb` methods/functions be added? For now, in my emulator, I've resorted to using only 2.88mb disks and swapping the buffers with the following: (assuming `windows2app` points to the Window of an `<iframe>`)
> 
> ```js
> windows2app.emulator.v86.cpu.devices.fda.insert_disk({buffer: buffer1})
> windows2app.emulator.v86.cpu.devices.fdb.insert_disk({buffer: buffer2})
> ```
> 
> The problem with this is only 2.88mb (or whatever I use by default) are supported, as in one size of disk only. Using `set_fda` allows me to use any size disk for FDA and swap sizes during usage. If a `set_fdb` existed, this would allow for the same multiple sizes without recreating the emulator object.

If there is no disk inserted during boot we need to fall back to a default drive type. Each floppy drive has a drive type that must be defined before we boot the guest. The guest (or BIOS) will usually read this drive type exactly once during boot.

For testing I still had fdb default to a 2.88M and fda to a 1.44M drive, I changed it and now both drives default to 1.44M (3"1/2) which I believe is most universal.

Note that if there is a disk inserted before booting its maximum disk size is 3.84 MB (3932160 bytes).

I think this fixes your issue, can you confirm?

chschnell commented:
The code of this PR is ok, far from perfect, but it has a BSD/GPL license conflict.

I'm not able to base this PR on the floppy controller documentation available to me (for a clean-room implementation it would likely be neccessary to have access to an actual physical device).

The best (only, actually) reference I was able to find is qemu, it is the only source for some internal controller state change sequences and two important data structures that I need. This PR implements about half of the stuff that qemu has, meaning some parts of it are a subset of qemu.

In some instances I don't have an alternative to source the required information from.

I can't decide what outweighs what here. I'm sorry for having created this situation, if I'd known in advance that qemu is excluded I wouldn't have written this PR in the first place. It wouldn't be a problem for me if this PR gets rejected, it might still be usefull as a reference for any future efforts.

@copy: Any ideas on how to proceed with this?

copy commented:
> @copy: Any ideas on how to proceed with this?

I just had a quick look at fdc.c from qemu, and it actually has an MIT license header. I think we can make this work:

- Copy the license header from fdc.c into LICENSE.MIT
- Mention at the top of floppy.js: `// this file is licensed under both BSD and MIT, see LICENSE and LICENSE.MIT`
- Mention in the readme at the bottom of license section:

> src/floppy.js contains parts ported from qemu under the MIT license, see LICENSE.MIT

Let me know if think this is ready for a code-review.

chschnell commented:
Ok, we're ready to go here.

The final step was to integrate the revised FloppyController into the v86 API and web UI.

Changes to the web UI (index.html and debug.html):
- added fdb image selector
- added empty disk options for fda and fdb (constrained to 160..3840 KB, default: 1440 KB)
- added insert/eject button for fdb

Changes to the V86 API:

```javascript
V86.set_fda(file)     // as before: insert disk into fda
V86.set_fdb(file)     // new
V86.eject_fda()       // as before: eject disk from fda
V86.eject_fdb()       // new
V86.get_disk_fda()    // new, return fda disk buffer or null
V86.get_disk_fdb()    // new
```

The get_disk_fd(a|b) methods were added for convenience (the insert/eject floppy buttons need them).

chschnell commented on LICENSE.MIT:
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2012, The v86 contributors
```
Done, this can be resolved.

chschnell commented on src/floppy.js:
```diff
@@ -33,534 +228,1288 @@ export function FloppyController(cpu, fda_image, fdb_image)
+    this.perpendicular = 0;
+    this.read_only = false;
+    this.media_changed = true;
+    this.img_buffer = null;
```
Done, this can be resolved.

chschnell commented on src/floppy.js:
```diff
@@ -33,534 +228,1288 @@ export function FloppyController(cpu, fda_image, fdb_image)
+    /** @const */
+    this.fdc = fdc;
+    /** @const */
+    this.cpu = fdc.cpu;
```
Done, this can be resolved.

chschnell commented:
> > And this simple test also freezes in the last step, so it does not appear to be related to the floppy device. When I ran it with log_level = 3, it continuously spammed to the console:
> 
> Looking at the screen, it looks like the keyboard text is cut off:
> 
> ```
> C:\>more < C:\V86TE  
> ```
> 
> I suspect `keyboard_send_text` is too fast and MS-DOS's keyboard buffer is only 16 bytes. Can you try with this:
> ```
> async function send_text(emulator, text)
> ```

That fixed it, thanks! I noticed that too and had the same idea, just hadn't come around to test it yet :)

> > First, I couldn't find a linux image that has floppy kernel support.
> 
> Here is one that should work: https://copy.sh/v86/images/Core-4.7.7.iso

I wrote a test with Core-4.7.7.iso and it also works here, however its image is not available on github and so the latest commit failed.

It also fails when I use the delayed keyboard function `send_text()`, I must use `emulator.keyboard_send_text();` directly here or the keyboard input turns all garbage:

```
 (þ-
 //\\   Core is distributed with ABSOLUTELY NO WARRANTY.
 v_/_           www.tinycorelinux.com

tc@box:~$ mslkusdd io-r l ma/o mu/nnmttn/ tf//ddfaed
-sh: mslkusdd: not found
tc@box:~$ va/
-sh: va/: not found
tc@box:~$ fd0 /mnt/fda
-sh: fd0: not found
tc@box:~$
```

Note that the actual keyboard input that I'm sending is:

```
mkdir /mnt/fda
sudo mount /dev/fd0 /mnt/fda
ls -la /mnt/fda
```

So the keyboard is a bit troubled and depends on the guest OS, however I just ran `make api-tests` in a bash for-loop for 10 times, it's robust here.

All 4 floppy tests that were suggested now work with the latest commit.

chschnell commented:
> If it will be useful for CI testing, I created FreeDOS images with a serial port console (with and without CD-ROM driver): [fdos-serial.tar.gz](https://github.com/user-attachments/files/21724365/fdos-serial.tar.gz).
> 
> The serial port in DOS is a bit of buggy thing, you need to press Enter on PS/2 keyboard after startup to enable input from serial port.
> 

Thanks, I will try it out.

> > First, I couldn't find a linux image that has floppy kernel support.
> 
> i.copy.sh/TinyCore-11.0.iso is available on CI and supports floppy drives, but there is no serial port console like in Buildroot.
> 

I can see that it detects both floppy drives, but it boots into a graphical Desktop here. Do you know an easy way to make it boot into the text console (I think I could make it through the kernel command line)?

> I also found a bunch of FreeDOS floppy images of various formats for testing: https://github.com/codercowboy/freedosbootdisks/tree/master/bootdisks.

I will test them all :)

SuperMaxusa commented:
> Do you know an easy way to make it boot into the text console (I think I could make it through the kernel command line)?

Select "Boot Core (command line only)" and press Enter:

```js
const CONFIG_CORE11_CD = {
    bios: { url: __dirname + "/../../bios/seabios.bin" },
    vga_bios: { url: __dirname + "/../../bios/vgabios.bin" },
    cdrom: { url: __dirname + "/../../images/TinyCore-11.0.iso" },
    autostart: true,
    memory_size: 128 * 1024 * 1024,
    log_level: 0,
    disable_jit: +process.env.DISABLE_JIT
};

await run_floppy_test("floppy-core11-linux", CONFIG_CORE11_CD, 90, async emulator =>
{
    console.log("Waiting for boot menu");
    await emulator.wait_until_vga_screen_contains("Boot media is removable. Use TAB to edit options for specific needs.");
    await pause(1500);

    // press arrow down key 3 times
    for(let i = 0; i < 3; i++)
    {
        emulator.keyboard_send_scancodes([
            0xe0, 0x50,        // press 
            0xe0, 0x50 | 0x80, // release 
        ]);
        await pause(600);
    }
    emulator.keyboard_send_text("\n");

    console.log("Waiting for tc@box:~$");
    await emulator.wait_until_vga_screen_contains("tc@box:~$ ");

    console.log("Inserting disk freedos722.img into drive A:");
    emulator.set_fda({ url: __dirname + "/../../images/freedos722.img" });
    await pause(1000);

    console.log("Mounting /dev/fd0 into /mnt/fda");
    emulator.keyboard_send_text("mkdir /mnt/fda\n");
    await emulator.wait_until_vga_screen_contains("tc@box:~$ ");
    emulator.keyboard_send_text("sudo mount /dev/fd0 /mnt/fda\n");
    await emulator.wait_until_vga_screen_contains("tc@box:~$ ");

    console.log("Reading /mnt/fda");
    emulator.keyboard_send_text("ls -la /mnt/fda\n");
    await emulator.wait_until_vga_screen_contains("x86test.asm");
});
```

chschnell commented on tests/api/floppy.js:
```diff
@@ -0,0 +1,170 @@
+    await emulator.wait_until_vga_screen_contains("C:\\> ");
+
+    console.log("Reading file A:\\V86TEST.TXT");
+    send_text(emulator, "more < A:\\V86TEST.TXT\n");
```
That's better and I changed it, thanks!

chschnell commented:
> > Do you know an easy way to make it boot into the text console (I think I could make it through the kernel command line)?
> 
> Select "Boot Core (command line only)" and press Enter:
> 

Awesome, I switched to TinyCore-11.0.iso and it works!

SuperMaxusa commented on tests/api/floppy.js:
```diff
@@ -0,0 +1,170 @@
+    cdrom: { url: __dirname + "/../../images/Core-4.7.7.iso" },
+    autostart: true,
+    memory_size: 32 * 1024 * 1024,
+    filesystem: {},
```
> What is it used for?

This is for 9pfs. Since Tiny Core and MS-DOS don't support it, so you can safely remove it.

> Also, I removed field network_relay_url: "<UNUSED>", is this correct or does it serve any purpose here?

IIRC this is a placholder for state files and when guest OS tries to start a DHCP client.

chschnell commented on src/floppy.js:
```diff
@@ -33,534 +228,1288 @@ export function FloppyController(cpu, fda_image, fdb_image)
+    // the disk image buffer, only a few register states, so a floppy drive
+    // remained essentially unchangd when a state snapshot was applied.
+    // The snapshotted registers can be safely ignored when restoring state,
+    // hence the entire old-style state is now ignored and deprecated.
```
I did now, here's the scenario I tested, first:

- booted https://copy.sh/v86/ into my 20m FreeDOS HD (old snapshot format)
- inserted floppy disk into fda
- excecuted CLI commands `cls` and `dir A:`
- saved state snapshot

then:

- booted my local V86 into the same 20m FreeDOS HD (new snapshot format)
- restored state snapshot from above
- excecuted `dir A:` again

and access to fda failed as expected:

<img width="720" height="400" alt="freedos-fd-old-new-state-snapshot" src="https://github.com/user-attachments/assets/1b270c45-a06f-4555-aff7-ac4f0803909b" />

SuperMaxusa commented:
> This test in tests/api/floppy.js is commented out for now, let me know if you want it in. If so, please upload the 8 images to github CI so I can activate the test (see the comment in tests/api/floppy.js).

Just an idea, but maybe we can move this test (https://github.com/copy/v86/pull/1368/commits/c3713faa3aba4aa2d588e05b4b72867830ab8d19) to [`tests/full/run.js`](https://github.com/copy/v86/blob/51bf5a63dcc34f806cdc3c182c383d770e76e0cf/tests/full/run.js)? It has an automatic test skip if the image wasn't found.

I think testing only "uncommon" (e.g. 180K, 320K, 360K, 640K etc) disk formats should be enough.

Disk formats that already exist in full test:
- 720K - `FreeDOS`
- 1.44M - `Windows 1.01`, `Sol OS`, `Snowdrop`, `KolibriOS`, `QNX`, `OpenBSD`, `Mobius`
- 2K - `ASM Space Invaders`

By the way, we have a 160K image of 86dos on the CDN, but it not used in tests, maybe it would be good to add it? i.copy.sh/pc86dos.img

copy commented on tests/api/floppy.js:
```diff
@@ -0,0 +1,326 @@
+ * @param {number} response_timeout_msec
+ * @return {Array[string]}
+ */
+async function expect(emulator, command, response_lines, response_timeout_msec)
```
This looks nice, but we already have two other similar implementations of "matching text on the vga screen":

- An older one in https://github.com/copy/v86/blob/51bf5a63dcc34f806cdc3c182c383d770e76e0cf/tests/full/run.js
- A newer, but slightly incomplete one: https://github.com/copy/v86/blob/51bf5a63dcc34f806cdc3c182c383d770e76e0cf/src/browser/starter.js#L1248

And they all work subtly different: The above two use `.startsWith` and `.contains` respectively, while this one uses `.endsWith` and matches only at the bottom.

I'd suggest the following way forward (but let me know if you disagree; since the tests already seem to work, we could merge as is):

- You merge your multi-line matching from this function into `V86.prototype.wait_until_vga_screen_contains` and call that instead
- `wait_until_vga_screen_contains` changes its signature to `expect: Array[string | RegExp] | string | RegExp, options?: { bottom: boolean }`, where `bottom` defaults to false and string matching uses `.startsWith` (I've had a quick look at the uses, and all of them expect matching at the beginning of the line). If you need matching at the end of the line, you can use a regexp with `$` 
- We leave tests/full/run.js untouched for now, and I'll look into it later

chschnell replied:
We could do without `options.bottom` if we said that this new multiline mode is active if `Array.isArray(expect)` is true. An array argument is not currently supported by `wait_until_vga_screen_contains()`.

What do you think?

chschnell replied:
All right, it's in. Merging the two functions worked out quite well, I believe, let me know if you see any issues with it. The new (backward compatible) prototype now is:

```javascript
 **/
 * @param {string|RegExp|Array<string|RegExp>} expected
 * @param {{timeout_msec:(number|undefined)}=} options
 */
V86.prototype.wait_until_vga_screen_contains = async function(expected, options)
```

The optional `options.timeout` can be used in both cases, and the function now also returns the captured VGA screen text.

What's left to do is to upload these images to the CDN [from here](https://github.com/codercowboy/freedosbootdisks/tree/master/bootdisks):

```
/images/freedos-fds/freedos.boot.disk.160K.img
/images/freedos-fds/freedos.boot.disk.180K.img
/images/freedos-fds/freedos.boot.disk.320K.img
/images/freedos-fds/freedos.boot.disk.360K.img
/images/freedos-fds/freedos.boot.disk.640K.img
/images/freedos-fds/freedos.boot.disk.1200K.img
```

I improvised the subdirectory `/images/freedos-fds` for this, let me know if you want these paths to be changed in `run.js`.

Maybe you saw it, `RegExp.escape()` is too new (May 2025 Baseline) for the current github CI setup. Please let me know once this function is supported, I'd like to use that instead of the helper function that I found [on stackoverflow](https://stackoverflow.com/a/3561711).

copy commented on tests/api/floppy.js:
```diff
@@ -0,0 +1,326 @@
+        {
+            // prepend command to expected response lines
+            response_lines = Array.from(response_lines);
+            response_lines.unshift(trimmed_command);
```
`[trimmed_command, ...response_lines]`

chschnell commented on tests/api/floppy.js:
```diff
@@ -0,0 +1,225 @@
+ */
+export async function expect(emulator, command, expected, timeout_msec)
+{
+    if(command)
+    {
+        for(const ch of command)
+        {
+            emulator.keyboard_send_text(ch);
+            await pause(10);
+        }
+        command = command.trimRight();
+        if(command)
+        {
+            expected = [new RegExp(regexp_escape(command) + "$"), ...expected];
+        }
+    }
```
Imagine the guest is in some interactive text menu (boot menu), we send command `" "` or `"\n"` and expect a list of lines which happen to sit at the top of the screen, we would now expect an extra line above the screen which would fail.

I'd argue that if we send either of these commands (specifically whitespace) we cannot detect them later on screen, that's why I disregard these entirely in the expected lines.

Definitely a corner-case, or is it an over-interpretation?

I kept it for now, just wanted to hear your opinion.

chschnell commented on src/browser/starter.js:
```diff
@@ -1299,52 +1358,126 @@ V86.prototype.automatically = function(steps)
+    }
+    finally
+    {
+        this.remove_listener("screen-put-char", screen_put_char);
```
I have removed the return value from `V86.wait_until_vga_screen_contains()`, it would now be possible to do away with the exception and make it return a boolean, `true` in case of success and `false` if it times out.

How about that?

copy replied:
Sounds good.

chschnell commented:
> I'm nitpicking at this stage. Feel free to ignore any comments.
> 

I don't mind at all.

The last commit 72d66c4 resolves all things that you brought up.

As a matter of experience I'm really strict about automated testing, I know.

> From my point of view this is good to merge. Last question, would you like to squash or merge as-is? (the commit history looks good, so I'd merge without squash)

If you're still fine with the history then let's merge it.

Thanks for all your feedback!
